zerver/lib/upload.py

```diff
@@ -322,33 +311,51 @@ def get_file_info(request: HttpRequest, user_file: File) -> Tuple[str, int, Opti
 
 
 def get_signed_upload_url(path: str) -> str:
-    conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
-    return conn.generate_url(15, 'GET', bucket=settings.S3_AUTH_UPLOADS_BUCKET, key=path)
+    client = boto3.client(
+        's3',
+        aws_access_key_id=settings.S3_KEY,
+        aws_secret_access_key=settings.S3_SECRET_KEY)
+    params = {
+        'Bucket': settings.S3_AUTH_UPLOADS_BUCKET,
+        'Key': path}
+    return client.generate_presigned_url(
+        ClientMethod='get_object',
+        Params=params,
+        ExpiresIn=15,
+        HttpMethod='GET')
 
 def get_realm_for_filename(path: str) -> Optional[int]:
-    conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
-    key = get_bucket(conn, settings.S3_AUTH_UPLOADS_BUCKET).get_key(path)  # type: Optional[Key]
-    if key is None:
-        # This happens if the key does not exist.
+    try:
+        s3_object = get_s3_object(settings.S3_AUTH_UPLOADS_BUCKET, path)
+        user_profile_id = s3_object.metadata['user_profile_id']
+    except Exception:
+        # Boto3 has weird exception handling that doesn't
+        # work that reliably
+        # https://github.com/boto/boto3/issues/1195
         return None
-    return get_user_profile_by_id(key.metadata["user_profile_id"]).realm_id
+    return get_user_profile_by_id(user_profile_id).realm_id
 
 class S3UploadBackend(ZulipUploadBackend):
     def __init__(self) -> None:
-        self.connection = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        self.session = boto3.Session(settings.S3_KEY, settings.S3_SECRET_KEY,
+                                     region_name='s3.amazonaws.com')
 
     def delete_file_from_s3(self, path_id: str, bucket_name: str) -> bool:
-        bucket = get_bucket(self.connection, bucket_name)
-
-        # check if file exists
-        key = bucket.get_key(path_id)  # type: Optional[Key]
-        if key is not None:
-            bucket.delete_key(key)
-            return True
-
-        file_name = path_id.split("/")[-1]
-        logging.warning("%s does not exist. Its entry in the database will be removed." % (file_name,))
-        return False
+        try:
+            s3_object = get_s3_object(bucket_name, path_id)
+            # Call `HEAD object` to see if the object exists
+            s3_object.load()
+        except Exception:
```

Comment:
So sometimes we see a `ClientError`, other times we might see a `NoSuchBucket` error.  However actually  specifying the exception is super weird as explained in the GH issue mentioned.  Not sure of the best way to deal with this.